### PR TITLE
Be nc news 3

### PR DESCRIPTION
**Pull request by dancancode85**

### CODE REVIEW

File: __tests__/integration.test.js [hunk 1]
```diff
@@ -1,30 +1,102 @@
...
+    return request(app)
+      .get("/api")
+      .expect(200)
+      .then((res) => {
+        const testObject = endpoints;
+        const endpointResult = res.body;
+        expect(testObject).toEqual(endpointResult);
+      });
+  });
+});
```
Comment (MKD01) [inline __tests__/integration.test.js:45]: Really good test

File: __tests__/integration.test.js [hunk 2]
```diff
@@ -1,30 +1,102 @@
...
+
+describe("getArticleByID", () => {
+  it("should return correct article", () => {
+    return request(app)
+      .get("/api/articles/1")
+      .expect(200)
+      .then((res) => {
+        const testArticle = res.body.article[0];
```
Comment (MKD01) [inline __tests__/integration.test.js]: We should not be expecting an array here, res.body.article should be the single article object

File: __tests__/integration.test.js [hunk 3]
```diff
@@ -1,30 +1,102 @@
...
+// title
+// article_id
+// body
+// topic
+// created_at
+// votes
+// article_img_url
+// Consider what errors could occur with this endpoint, and make sure to test for them.
+
+// Remember to add a description of this endpoint to your /api endpoint.
```
Comment (MKD01) [inline __tests__/integration.test.js]: Dont forget to remove comments

File: __tests__/integration.test.js [hunk 4]
```diff
@@ -1,30 +1,102 @@
...
+      .then((res) => {
+        expect(res.body.error).toBe("Article id 999 not found");
+      });
+  });
+  it("should return a 400 error if invalid ID passed ", () => {
+    return request(app)
+      .get("/api/articles/dog")
+      .expect(400)
+      .then((res) => {
+        expect(res.body.error).toBe("Bad request, dog is not a valid request");
```
Comment (MKD01) [inline __tests__/integration.test.js]: There is no need to have specific error messages, just a generic bad request message is enough

File: __tests__/integration.test.js [hunk 5]
```diff
@@ -1,30 +1,102 @@
...
+        };
+        expect(testArticle).toEqual(expectedArticle);
+      });
+  });
+  it("should return a 404 error if no such id exists ", () => {
+    return request(app)
+      .get("/api/articles/999")
+      .expect(404)
+      .then((res) => {
+        expect(res.body.error).toBe("Article id 999 not found");
```
Comment (MKD01) [inline __tests__/integration.test.js]: There is no need to have very specific error messages, just a generic not found message is enough

File: app.js [hunk 6]
```diff
@@ -1,7 +1,23 @@
...
+//   } = require("./errors");
+
+app.get("/api/topics", getTopics);
+app.get("/api", getAPI);
+app.get("/api/articles/:id", getArticleByID);
+
+// app.use(handlePSQLErrors);
+// app.use(handleCustomErrors);
+// app.use(handleServerErrors);
+
```
Comment (MKD01) [inline app.js]: Im not sure why there is this commented out code here

File: controllers/controller.js [hunk 7]
```diff
@@ -1,10 +1,47 @@
...
-  })
-};
\ No newline at end of file
+  getAllTopics()
+    .then((rows) => {
+      res.status(200).send({ topics: rows });
+    })
+    .catch((err) => {
+      res.status(err.status).send({ msg: err.message });
+    });
```
Comment (MKD01) [inline controllers/controller.js]: These errors should ideally be extracted out from here and handled in some error handling middleware functions

File: controllers/controller.js [hunk 8]
```diff
@@ -1,10 +1,47 @@
...
+};
+
+exports.getAPI = (req, res, next) => {
+  res.status(200).send(endpoints);
+};
+
+exports.getArticleByID = (req, res, next) => {
+  const { id } = req.params;
+  console.log(id);
```
Comment (MKD01) [inline controllers/controller.js]: Dont forget to remove console.logs

File: controllers/controller.js [hunk 9]
```diff
@@ -1,10 +1,47 @@
...
+  getArticlesByID(id)
+    .then((article) => {
+      console.log(article);
```
Comment (MKD01) [inline controllers/controller.js]: And here

File: controllers/controller.js [hunk 10]
```diff
@@ -1,10 +1,47 @@
...
+      if (!article || article.length === 0) {
+        res.status(404).send({ error: `Article id ${id} not found` });
+      } else {
```
Comment (MKD01) [inline controllers/controller.js]: This shouldnt be handled in the controller

File: controllers/controller.js [hunk 11]
```diff
@@ -1,10 +1,47 @@
...
+      //    :res.status(200).send({article: data})
+    })
+    .catch((err) => {
+      if (!id || isNaN(+id)) {
+        res
+          .status(400)
+          .send({ error: `Bad request, ${id} is not a valid request` });
+      } else {
+        res.status(500).send({ error: "Internal server error" });
+      }
```
Comment (MKD01) [inline controllers/controller.js]: Same as the above this shouldnt be handled here, but instead in some error handling middleware functions.

File: endpoints.json [hunk 12]
```diff
@@ -25,5 +25,31 @@
...
+    "possibleErrors": [
+      {
+        "status": 404,
+        "description": "Article not found"
+      },
+      {
+        "status": 500,
+        "description": "Internal server error"
+      }
+    ]
```
Comment (MKD01) [inline endpoints.json]: This is an interesting idea, but not something a user should need to know. We should expect the user to only see the valid ways to use this api in this json file so it should not contain any errors here.

File: error-handling.md [hunk 13]
```diff
@@ -33,43 +33,43 @@ The following is _not_ a comprehensive list! Its purpose is just to get the ball
...
+404 Not Found - Well formed `article_id` that doesn't exist in the database (e.g. `/999999`)
 
 ### PATCH `/api/articles/:article_id`
 
-- Bad `article_id` (e.g. `/dog`)
-- Well formed `article_id` that doesn't exist in the database (e.g. `/999999`)
-- Invalid `inc_votes` (e.g. `{ inc_votes : "cat" }`)
+400 Bad Request- Bad `article_id` (e.g. `/dog`)
+404 Not Found - Well formed `article_id` that doesn't exist in the database (e.g. `/999999`)
+422 Unprocessable Entity - Invalid `inc_votes` (e.g. `{ inc_votes : "cat" }`)
```
Comment (MKD01) [inline error-handling.md:51]: Im not sure what the need for this file is

File: models/models.js [hunk 14]
```diff
@@ -1,12 +1,31 @@
...
+  });
+};
+
+const getAllEndpoints = () => {
+  return db.query(`SELECT * FROM api;`).then((data) => {
+    return data.rows;
+  });
+};
+const getArticlesByID = (id) => {
+  console.log("Received ID:", id);
```
Comment (MKD01) [inline models/models.js]: Dont forget to remove console.logs

**Review by MKD01 — CHANGES_REQUESTED** [review]
Good work so far, but there are some changes I would like to see made before you merge this in, so please re-submit this pr once you have made the requested changes 👍

**Review by nc-hev-mag — APPROVED** [review]
Looks good to merge based on suggested changes!

**Pull request merged**